# 220968280 - Anirudh Swaminathan - DSE A2

## week 1 - exercise 2

In [1]:
!pip3 -V

pip 22.3.1 from /home/mca/Documents/220968280/venv/lib/python3.11/site-packages/pip (python 3.11)


In [2]:
import pandas as pd
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
import matplotlib.pyplot as plt
%matplotlib inline

2024-07-24 09:47:14.145316: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-07-24 09:47:14.174154: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-24 09:47:14.624174: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [3]:
df = pd.read_csv("/home/mca/Documents/220968280/week1/bodyfat.csv")

In [4]:
df

,Density,BodyFat,Age,Weight,Height,Neck,Chest,Abdomen,Hip,Thigh,Knee,Ankle,Biceps,Forearm,Wrist
0,1.0708,12.3,23,154.25,67.75,36.2,93.1,85.2,94.5,59.0,37.3,21.9,32.0,27.4,17.1
1,1.0853,6.1,22,173.25,72.25,38.5,93.6,83.0,98.7,58.7,37.3,23.4,30.5,28.9,18.2
2,1.0414,25.3,22,154.00,66.25,34.0,95.8,87.9,99.2,59.6,38.9,24.0,28.8,25.2,16.6
3,1.0751,10.4,26,184.75,72.25,37.4,101.8,86.4,101.2,60.1,37.3,22.8,32.4,29.4,18.2
4,1.0340,28.7,24,184.25,71.25,34.4,97.3,100.0,101.9,63.2,42.2,24.0,32.2,27.7,17.7
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
247,1.0736,11.0,70,134.25,67.00,34.9,89.2,83.6,88.8,49.6,34.8,21.5,25.6,25.7,18.5
248,1.0236,33.6,72,201.00,69.75,40.9,108.5,105.0,104.5,59.6,40.8,23.2,35.2,28.6,20.1
249,1.0328,29.3,72,186.75,66.00,38.9,111.1,111.5,101.7,60.3,37.3,21.5,31.3,27.2,18.0
250,1.0399,26.0,72,190.75,70.50,38.9,108.3,101.3,97.8,56.0,41.6,22.7,30.5,29.4,19.8


In [5]:
num_samples = 252
num_features = 14
X = df.drop(["BodyFat"], axis=1)
y = df['BodyFat']

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

In [7]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [8]:
model = Sequential([
    Input(shape=(num_features,), name='input'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(1, activation='relu')
])

2024-07-24 09:47:15.548427: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-07-24 09:47:15.578366: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-07-24 09:47:15.578520: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysf

In [9]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 128)               1920      
                                                                 
 dense_1 (Dense)             (None, 64)                8256      
                                                                 
 dense_2 (Dense)             (None, 32)                2080      
                                                                 
 dense_3 (Dense)             (None, 1)                 33        
                                                                 
Total params: 12289 (48.00 KB)
Trainable params: 12289 (48.00 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


Model summary: Shows our 3 hidden layers of 128, 64 and 32 neurons each with a single neuron in the output layer. Number of parameters in each hidden layer is 1920, 8256 and 2080 respectively. We are updating all 12289 parameters here, which is why we have 0 non-trainable parameters.

In [10]:
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.1), loss='MeanSquaredError')

In [11]:
epochs = 50
batch_size = 32
h = model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, validation_split=0.1)

Epoch 1/50


2024-07-24 09:47:17.061712: I tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:606] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2024-07-24 09:47:17.063633: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7f6e0c2da530 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2024-07-24 09:47:17.063647: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): NVIDIA GeForce RTX 3060, Compute Capability 8.6
2024-07-24 09:47:17.066799: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:255] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2024-07-24 09:47:17.161757: I tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:432] Loaded cuDNN version 8600
2024-07-24 09:47:17.234051: I ./tensorflow/compiler/jit/device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the p

7/7 [==============================] - 2s 19ms/step - loss: 3244.6707 - val_loss: 318.7419
Epoch 2/50
7/7 [==============================] - 0s 4ms/step - loss: 269.5108 - val_loss: 260.6645
Epoch 3/50
7/7 [==============================] - 0s 4ms/step - loss: 404.5843 - val_loss: 393.0221
Epoch 4/50
7/7 [==============================] - 0s 4ms/step - loss: 433.5731 - val_loss: 357.2997
Epoch 5/50
7/7 [==============================] - 0s 6ms/step - loss: 333.3631 - val_loss: 310.5866
Epoch 6/50
7/7 [==============================] - 0s 5ms/step - loss: 253.3329 - val_loss: 178.0307
Epoch 7/50
7/7 [==============================] - 0s 3ms/step - loss: 151.3268 - val_loss: 139.5824
Epoch 8/50
7/7 [==============================] - 0s 5ms/step - loss: 170.8989 - val_loss: 150.1119
Epoch 9/50
7/7 [==============================] - 0s 4ms/step - loss: 104.5609 - val_loss: 72.7511
Epoch 10/50
7/7 [==============================] - 0s 3ms/step - loss: 68.0776 - val_loss: 63.3523
Epoch 11/50

In [12]:
loss = model.evaluate(X_test, y_test, batch_size=batch_size)
print("Test Loss:", loss)

1/1 [==============================] - 0s 13ms/step - loss: 5.7800
Test Loss: 5.779993057250977
